students/mgummel/lesson05/mailroom3.py

```diff
@@ -0,0 +1,231 @@
+        total_donation, avg_donation, num_of_donations = compute_donation(
+            donor_db[k])
+        computed_list.append(
+            (k, total_donation, num_of_donations, avg_donation))
+    sorted_list = sorted(computed_list, key=itemgetter(1), reverse=True)
+
+    for sorted_item in sorted_list:
+        print("{:<27}${:>12.2f}{:>14}   ${:>13.2f}".format(*sorted_item))
+
+
+def create_header():
+    """
+    Prints a table in a nice formatted way that contains the 
+    table/titles that will organize  donor name, the total amount 
+    they have donated, the number of donations, and their 
+    corresponding average gift amount in a neat formatted manner.
+    """
+
+    header = "{:<26}|{:^15}|{:^13}|{:>14}".format(
+        "Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    print("\n" + header)
+    print("-" * len(header))
+
+
+def compute_donation(list_of_donations):
+    """
+    Accepts a list and calculates the total amount given,
+    the number of donations, and the average amount of each
+    donor's donation.
+
+    :param donations: list of donations per donor
+    :type donations: list
+    """
+
+    total_amount = 0
+    donation_num = 0
+
+    # Increment both number of donations and the total donation amount
+    for amount in list_of_donations:
+        donation_num += 1
+        total_amount += amount
+
+    total_amount = round(total_amount, 2)
+    avg_amount = round(total_amount / donation_num, 2)
+
+    return total_amount, avg_amount, donation_num
+
+
+def write_letter():
+    translate_dict = {ord(" "): "_", ord(","): None}
+    directory = pathlib.Path("")
+
+    while True:
+            location = input("\n".join(("Please type the directory you want files to be generated.",
+                                        "(Hit \'Enter\' to default to the current working directory)",
+                                        ">>> ")))
+            if location == "quit":
+                quit_program()
+            directory = pathlib.Path(location)
+            if directory.exists():
+                break
+            else:
+                print("That's not a valid directory. Please try again.")
+                directory = pathlib.Path("")
+
+    for key_item in donor_db.keys():
+        filepath = directory / key_item.translate(translate_dict)
+
+        with open(f'{filepath}.txt', 'w', encoding='utf-8') as email:
+            email_template = build_template(key_item, donor_db[key_item])
+            email.write(email_template)
+           
+
+def main():
+    menu = {
+        "1": send_thank_you,
+        "2": create_report,
+        "3": write_letter,
+        "4": quit_program
+    }
+
+    print("Welcome to the Mailroom App!")
+
+    while True:
+        response = input(prompt)
+        if response in menu:
+            menu[response]()
+            print("\n")
+        else:
+            print("Not a valid option!\n\n")
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
@mgummel25 - I'm not seeing exceptions or comprehensions in this file as required by the assignment.  Let me know if you need help.

Reply:
@charlesrobison I'm still working on it. Will upload the latest version ASAP.